File: exercise/python/dropbox_tools.py
"""
dropbox_tools.py: methods for performing some Dropbox actions (creating or printing files etc)

You should import this file as follows so you can call its methods directly:
    from dropbox_tools import *

Dependencies:
    python 2.7
    dropbox - Python Dropbox API (https://www.dropbox.com/developers/datastore/sdks/python)

Global constants:
    ACCOUNT_INFO_FILE: This file is created in the Dropbox app folder once the client has authorised with Dropbox
        It contains information on the Dropbox user account
    REVIEW_DIRECTORY: This folder is created in the Dropbox app folder once the client has authorised with Dropbox
    REVIEW_FILE: This file is created in the Dropbox app folder once the client has authorised with Dropbox

Global variables:
    logger: used to log error, info and debug messages
        example usage: logger = CO.logger; logger('message')

Methods:
    db_list_directory(): list the contents of the given Dropbox directory
    db_print_file(): print the contents of the given file to stdout
    db_create_text_file(): create a file and write some lines of text to it
    db_create_sample_files(): create some files and directories in Dropbox directory

"""

from tempfile import NamedTemporaryFile
import dropbox

import common_oauth as CO
logger = CO.logger

ACCOUNT_INFO_FILE = '/account_info.python.txt'
REVIEW_DIRECTORY = '/oauth_session_python'
REVIEW_FILE = '%s/oauth_session_review.python.md' % REVIEW_DIRECTORY

def db_list_directory(path='/'):
    """list the contents of the given Dropbox directory"""
    # INSERT CODE HERE:
    #  - create a Dropbox OAuth client object with which to make Dropbox calls
    #    hint: class is dropbox.client.DropboxClient()
    #  - this needs to be supplied with the access token
    #    hint: create a CO.AccessData() object and load() it from file

    # INSERT CODE HERE:
    #  - make a Dropbox client call to get the dropbox directory contents
    #    hint: Dropbox calls this "folder metadata"
    #    hint: @see https://www.dropbox.com/developers/core/docs/python (search for metadata())

    print 'CONTENTS OF %s:' % (file_path,)
    line_count = 1
    with client.get_file(file_path) as f:
        for line in f.readlines():
            print '{num:>2}: {line}'.format(num=line_count, line=line.strip('\n'))
            line_count += 1

def db_create_text_file(file_path='', lines=[]):
    """create a file and write some lines of text to it from stdin or lines[]"""
    # INSERT CODE HERE:
    #  - create a Dropbox OAuth client object (same code as above)

    if file_path == '':
        file_path = raw_input('Enter the name of a file to create: ').strip()
    tempfile = NamedTemporaryFile(delete=False)
    fname = tempfile.name
    if len(lines) ==0:
        while True:
            line = raw_input('Enter some text (blank to finish): ').strip('\n')
            if line.strip() == '': break
            tempfile.write('%s\n' % (line,))
    else:
        for line in lines:
            tempfile.write('%s\n' % (line,))
    tempfile.close()
    logger.debug('creating file %s...' % file_path)

    # INSERT CODE HERE:
    #  - open() the temporary file fname for read
    #  - pass this file handle to Dropbox put_file() to upload the file
    #    hint: @see https://www.dropbox.com/developers/core/docs/python (search for put_file())

def db_delete_file(file_path=''):
    if file_path == '':
        file_path = raw_input('Enter the name of the file to delete: ').strip()
    # INSERT CODE HERE:
    #  - create a Dropbox OAuth client object (same code as above)

    client.file_delete(file_path)
    logger.debug('deleted file %s' % file_path)

def db_disable_access_token():
    # INSERT CODE HERE:
    #  - create a Dropbox OAuth client object (same code as above)

    client.disable_access_token()
    logger.info('disabled Dropbox access token (access file not deleted)')

def db_create_sample_files():
    """create some files and directories in Dropbox directory"""
    # INSERT CODE HERE:
    #  - create a Dropbox OAuth client object (same code as above)


    # save file containing account information
    tempfile = NamedTemporaryFile(delete=False)
    fname = tempfile.name
    logger.debug('writing account info to temporary file %s' % (fname))
    tempfile.write('ACCOUNT INFORMATION FOR USER %s\n' % (access_data.user_id))
    account_info = client.account_info()
    for info in account_info:
        tempfile.write("{parameter} = {value}\n".format(parameter=info, value=str(account_info[info])))
    tempfile.close()
    tempfile = open(fname)
    logger.debug('uploading account info file %s' % ACCOUNT_INFO_FILE)
    client.put_file(ACCOUNT_INFO_FILE, tempfile, overwrite=True)
    logger.info('uploaded account info file %s' % ACCOUNT_INFO_FILE)

    # create review subdirectory
    try:
        client.file_delete(REVIEW_DIRECTORY)
    except dropbox.rest.ErrorResponse as e:
        if e.status <> 404: raise e
    client.file_create_folder(REVIEW_DIRECTORY)
    logger.info('created review directory %s' % REVIEW_DIRECTORY)

    # save file containing session review
    tempfile = NamedTemporaryFile(delete=False)
    fname = tempfile.name
    logger.debug('writing session review to temporary file %s' % (fname))
    tempfile.write('# KEEPING PASSWORDS PRIVATE WITH OAUTH\n')
    tempfile.write('An interesting and thought-provoking session\n')
    tempfile.write('The presenters were top-notch and I learned a lot\n')
    tempfile.close()
    tempfile = open(fname)

    logger.debug('uploaded session review file %s' % REVIEW_FILE)
    client.put_file(REVIEW_FILE, tempfile, overwrite=True)
    logger.info('uploaded session review file %s' % REVIEW_FILE)

File: exercise/java/src/oauth_demo/DropboxTools.java
package oauth_demo;

import java.util.Locale;
import java.util.Iterator;
import java.io.*;

import com.dropbox.core.*;

/**
 * Static methods for performing some Dropbox actions (creating or printing files etc).
 *
 * <p><b>Note</b>.
 * All paths in these methods are relative to the Dropbox app directory (eg
 * {@code ~/Dropbox/apps/bcs_spa_2014)}.
 * Dropbox paths always use forward-slashes, even on Windows.
 * <p>So, for example, to print the contents of {@code ~/Dropbox/apps/bcs_spa_2014/myfolder},
 * specify {@code myfolder} as a parameter ({@code /myfolder} will also work).
 *
 */
public class DropboxTools {

    /**
     * This file is created in the Dropbox app folder once the client has authorised with Dropbox.
     * It contains information on the Dropbox user account.
     * It is a Dropbox file path so uses forward-slash as directory separator, even on Windows.
     */
    static String DB_ACCOUNT_INFO_FILE = "/account_info.java.txt";

    /**
     * This folder is created in the Dropbox app folder once the client has authorised with Dropbox.
     * It is a Dropbox file path so uses forward-slash as directory separator, even on Windows.
     */
    static String DB_REVIEW_DIRECTORY = "/oauth_session_java";

    /**
     * This file is created in the Dropbox app folder once the client has authorised with Dropbox.
     * It is a Dropbox file path so uses forward-slash as directory separator, even on Windows.
     */
    static String DB_REVIEW_FILE = String.format("%s/oauth_session_review.java.md", DB_REVIEW_DIRECTORY);

    /** we don't ever instantiate this class but just call its methods statically */
    private DropboxTools() {};

    /**
     * Print the contents of a directory in the Dropbox app folder (using Dropbox calls, not file I/O).
     *
     * @param directory Dropbox path of the directory to print
     *
     * @throws IOException if there is an error creating the file
     * @throws DbxException if a Dropbox error occurs
     */
    public static void printDirectoryContents(String directory) throws IOException, DbxException {
        ConsoleLogger.debug("requested directory is %s", directory);
        String output = "\nCONTENTS OF DROPBOX DIRECTORY '"+directory+"': ";
        // EXERCISE:
        //  - create a Dropbox client
        //    hint: a helper function to create the Dropbox client is at the end of this file
        //  - create a Dropbox metada object which you will use to get the directory contents
        //    hint: the Dropbox class is DbxEntry.WithChildren
        //    @see http://dropbox.github.io/dropbox-sdk-java/api-docs/v1.7.x/com/dropbox/core/DbxEntry.WithChildren.html
// TODO ==> INSERT CODE HERE <==
        if (metadata == null) {
            output += "\n<none>";
        }
        else {
            // EXERCISE:
            //  - create a Dropbox metada iterator object which you will use to iterate over the sub-directories of this one
            //    hint: you need to iterate over metadata.children (which are all of type DbxEntry)
            //    @see http://dropbox.github.io/dropbox-sdk-java/api-docs/v1.7.x/com/dropbox/core/DbxEntry.WithChildren.html
// TODO ==> INSERT CODE HERE <==
            while (directoryIterator.hasNext()) {
                DbxEntry nextEntry = directoryIterator.next();
                String line = "";
                if (nextEntry instanceof DbxEntry.Folder) {
                    DbxEntry.Folder folderEntry = (DbxEntry.Folder)nextEntry;
                    output += String.format("\n %30s   %s%s", " ", folderEntry.name, "/");
                }
            }
            // EXERCISE:
            //  - create a Dropbox metada iterator object which you will use to iterate over the files in this directory
            //    hint: same as above
            directoryIterator = metadata.children.iterator();
            // SPA14_OAUTH_FINISH
            while (directoryIterator.hasNext()) {
                DbxEntry nextEntry = directoryIterator.next();
                String line = "";
                if (nextEntry instanceof DbxEntry.File) {
                    DbxEntry.File fileEntry = (DbxEntry.File)nextEntry;
                    output += String.format("\n %30s   %s (%s)", fileEntry.lastModified, fileEntry.name, fileEntry.humanSize);
                }
            }
        }
        System.out.println(output);
    }

    /**
     * Print the contents of a file in the Dropbox app folder (using Dropbox calls, not file I/O).
     *
     * @param printpath Dropbox path of the file to print
     *
     * @throws IOException if there is an error creating temporary files
     * @throws DbxException if a Dropbox error occurs
     */
    public static void printFile(String printpath) throws IOException, DbxException {
        // EXERCISE:
        //  - create a Dropbox client
        //    hint: a helper function to create the Dropbox client is at the end of this file
        //  - get the Dropbox file at printpath using client.getFile()
        //    hint: you need to create a temporary file into which you will download the Dropbox file
        //    @see http://dropbox.github.io/dropbox-sdk-java/api-docs/v1.7.x/com/dropbox/core/DbxClient.html#getFile(java.lang.String, java.lang.String, java.io.OutputStream)
// TODO ==> INSERT CODE HERE <==

        ConsoleLogger.debug("saved file contents to temporary file '%s'", tmpfile.getPath());
        BufferedReader br = new BufferedReader(new FileReader(tmpfile.getPath()));
        int lineCount = 1;
        String line = null;
        System.out.println("\nCONTENTS OF DROPBOX FILE '"+printpath+"': ");
        while ((line = br.readLine()) != null) {
            String printLine = String.format("%3d: %s", lineCount, line);
            System.out.println(printLine);
            lineCount++;
        }
        br.close();
    }

    /**
     * Create a file in the Dropbox app folder (using Dropbox calls, not file I/O).
     *
     * @param createpath Dropbox path of the file to create (this will be created under the Dropbox app folder)
     *
     * @throws IOException if there is an error creating temporary files
     * @throws DbxException if a Dropbox error occurs
     */
    public static void createFile(String createpath) throws IOException, DbxException {
        String tmpFilePath = File.createTempFile(AppData.APP_NAME, "txt").getPath();
        BufferedWriter bw = new BufferedWriter(new FileWriter(tmpFilePath));
        System.out.println("Enter file lines, blank line to finish");
        int lineCount = 0;
        while(true) {
            String line = System.console().readLine();
            if (line.length() == 0) { break; }
            bw.write(line+"\n");
            lineCount++;
        }
        bw.close();
        ConsoleLogger.debug("wrote %d lines to temporary file '%s'", lineCount, tmpFilePath);
        // EXERCISE:
        //  - create a Dropbox client
        //    hint: a helper function to create the Dropbox client is at the end of this file
        //  - upload the temporary file you created above into the Dropbox location createpath
        //    hint: use Dropbox force mode to overwrite the file if it exists
// TODO ==> INSERT CODE HERE <==
        ConsoleLogger.debug("copied temporary file '%s' to Dropbox path '%s'", tmpFilePath, createpath);
        f.delete(); 
    }

    /**
     * Create a file in the Dropbox app folder (using Dropbox calls, not file I/O).
     *
     * @param deletepath Dropbox path of the file to create (this will be created under the Dropbox app folder)
     *
     * @throws IOException if there is an error creating temporary files
     * @throws DbxException if a Dropbox error occurs
     */
    public static void deleteFile(String deletepath) throws IOException, DbxException {
        createDbxClient();
        client.delete(deletepath);
    }

    /**
     * Create some sample files in the Dropbox app folder (using Dropbox calls, not file I/O).
     *
     * <p>The method creates the following files:
     * <ul>
     *    <li>{@code DB_ACCOUNT_INFO_FILE} - contains information on the Dropbox user account
     *    <li>{@code DB_REVIEW_DIRECTORY} - a directory for DB_REVIEW_FILE
     *    <li>{@code DB_REVIEW_FILE} - contains a review of the session
     * </ul>
     *
     * @throws IOException if there is an error creating temporary files
     * @throws DbxException if a Dropbox error occurs
     */
    public static void createSampleFiles() throws IOException, DbxException {
        createDbxClient();
        DbxAccountInfo clientInfo = client.getAccountInfo();
        // create account info file
        String tmpFilePath = File.createTempFile(AppData.APP_NAME, "txt").getPath();
        BufferedWriter bw = new BufferedWriter(new FileWriter(tmpFilePath));
        bw.write("DROPBOX ACCOUNT INFORMATION\n");
        bw.write(String.format("country = %s\n", clientInfo.country));
        bw.write(String.format("displayName = %s\n", clientInfo.displayName));
        bw.write(String.format("referralLink = %s\n", clientInfo.referralLink));
        DbxAccountInfo.Quota quota = clientInfo.quota;
        bw.write(String.format("quota normal = %d, shared=%d, total=%d\n", quota.normal, quota.shared, quota.total));
        bw.close();
        ConsoleLogger.debug("wrote temporary account info file '%s'", tmpFilePath);
        File f = new File(tmpFilePath);
        ConsoleLogger.debug("uploading temporary account info file '%s' to Dropbox path '%s'", tmpFilePath, DB_ACCOUNT_INFO_FILE);
        client.uploadFile(DB_ACCOUNT_INFO_FILE, DbxWriteMode.force(), f.length(), new FileInputStream(f));
        ConsoleLogger.info("uploaded account info file '%s' to Dropbox path '%s'", tmpFilePath, DB_ACCOUNT_INFO_FILE);
        f.delete();
        // create Dropbox review directory if it doesn't already exist
        DbxEntry.Folder folderMetatada = client.createFolder(DB_REVIEW_DIRECTORY);
        if (folderMetatada == null) {
            ConsoleLogger.debug("review folder %s already exists", DB_REVIEW_DIRECTORY);
        }
        else {
            ConsoleLogger.debug("created review folder %s", DB_REVIEW_DIRECTORY);
        }
        // save file containing session review
        tmpFilePath = File.createTempFile(AppData.APP_NAME, "txt").getPath();
        bw = new BufferedWriter(new FileWriter(tmpFilePath));
        bw.write("# KEEPING PASSWORDS PRIVATE WITH OAUTH\n");
        bw.write("An interesting and thought-provoking session\n");
        bw.write("The presenters were top-notch and I learned a lot!\n");
        bw.close();
        ConsoleLogger.debug("wrote temporary review file '%s'", tmpFilePath);
        f = new File(tmpFilePath);
        ConsoleLogger.debug("uploading temporary review file '%s' to Dropbox path '%s'", tmpFilePath, DB_REVIEW_FILE);
        client.uploadFile(DB_REVIEW_FILE, DbxWriteMode.force(), f.length(), new FileInputStream(f));
        ConsoleLogger.info("uploaded review file '%s' to Dropbox path '%s'", tmpFilePath, DB_REVIEW_FILE);
    }

    /**
     * a Dropbox Dbx client variable used to perform Dropbox file actions.
     *
     * <p>Before using this client you must call {@code createDbxClient()}.
     *
     * @see <a href="http://dropbox.github.io/dropbox-sdk-java/api-docs/v1.7.x/com/dropbox/core/DbxClient.html">DbxClient reference</a>
     */
    public static DbxClient client = null;

    /**
     * Factory method to create a Dropbox DbxClient client.
     *
     * <p>The method performs the following steps:
     * <ol>
     * <li>Load the access token using {@code AccessData.load()}.
     * <li>Create a {@code DbxRequestConfig} object for the default locale.
     * <li>Create the {@code DbxClient} object.
     * </ol>
     *
     * @throws IOException if something bad has happened
     *
     */
    public static void createDbxClient() throws IOException {
        // EXERCISE:
        // - create a Dropbox OAuth client object with which to make Dropbox calls
        //   hint: class is DbxClient()
        // - this needs to be supplied with the access token
        //   hint: create an AccessData() object and load() it from file
        // - it also needs to be supplied with a Dropbox DbxRequestConfig() object
        //   hint: use the app name/version from AppData and the default Locale
        // - return this client to the caller
        // note how you have not had to provide any user credentials during this process!
// TODO ==> INSERT CODE HERE <==

        ConsoleLogger.debug("created DbxClient() object for config %s, access token %s", AppData.APP_NAME_VERSION, accessData.accessToken);
    }
}

